mono/metadata/threads.c

```diff
@@ -453,7 +454,44 @@ thread_get_tid (MonoInternalThread *thread)
 	return MONO_UINT_TO_NATIVE_THREAD_ID (thread->tid);
 }
 
-static void ensure_synch_cs_set (MonoInternalThread *thread)
+// Our problem with locking is complex.
+// We have managed threads. These managed threads have pointers
+// to managed objects of type mono_defaults.internal_thread_class. Our references below are
+// to these managed types. 
+//
+// We have a lot of racing requests to lock this thread. Crucially though, we need to be able
+// to free this lock while any of these other threads are waiting. 
+//
+// We have had a policy previously of re-initializing the mutex. This leads to a classic
+// partition usually seen in networks. The threads before the split will use one lock to
+// synchronize access to data, threads after will use another lock, and both will concurrently
+// dirty the same memory.
+//
+// Instead we needed to transition a thread from an initialized to a non-initializable state. Furthermore
+// we needed to ensure that locking attempts in progress finish before we free the lock. We do that by tracking
+// the number of waiters.
+//
+// Locker:
+// - inc num waiters
```

Comment:
Feels like a lock around the lock or such.

Reply:
More of a reference count

Reply:
But a count of the number of threads with incoming pointers, not the number of objects

Reply:
The comments here should help clarify that for future readers: https://github.com/mono/mono/pull/9486/files#diff-f1a9287559d3aa620b058619f47f00c9R584